runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java

```diff
@@ -143,13 +143,48 @@ public SimpleDoFnRunner(
   public void startBundle() {
     // This can contain user code. Wrap it in case it throws an exception.
     try {
-      invoker.invokeStartBundle(context);
+      invoker.invokeStartBundle(startBundleContext(context));
     } catch (Throwable t) {
       // Exception in user code.
       throw wrapUserCodeException(t);
     }
   }
 
+  private DoFn<InputT, OutputT>.Context startBundleContext(
+      final DoFn<InputT, OutputT>.Context delegate) {
+    return fn.new Context() {
+      @Override
+      public PipelineOptions getPipelineOptions() {
+        return delegate.getPipelineOptions();
+      }
+
+      @Override
+      public void output(OutputT output) {
+        throw new UnsupportedOperationException("Output is not supported within startBundle");
+      }
+
+      @Override
+      public void outputWithTimestamp(OutputT output, Instant timestamp) {
+        throw new UnsupportedOperationException("Output is not supported within startBundle");
+      }
+
+      @Override
+      protected Aggregator createAggregator(String name, CombineFn combiner) {
+        throw new UnsupportedOperationException();
+      }
+
+      @Override
+      public <T> void outputWithTimestamp(TupleTag<T> tag, T output, Instant timestamp) {
+        throw new UnsupportedOperationException("Output is not supported within startBundle");
+      }
+
+      @Override
+      public <T> void output(TupleTag<T> tag, T output) {
+        throw new UnsupportedOperationException("Output is not supported within startBundle");
```

Comment:
This is OK, but it might be even better and just as easy to split `Context` into `StartBundleContext` and `FinishBundleContext`.

Even better: Add automatic parameter support for `PipelineOptions` (if necessary - they should really be `ValueProvider` anyhow) and then forbid `Context` parameters via `DoFnSignatures`.